### PR TITLE
more improvements for E2EE.

### PR DESCRIPTION
**Pull request by cloudwebrtc**
* Use independent threads to process frame encryption/decryption
* Correct handle SIF frame, catch up with js rules
* Fix a fault tolerance judgment failure

### CODE REVIEW

File: api/crypto/frame_crypto_transformer.cc [hunk 1]
```diff
@@ -665,10 +680,11 @@ void FrameCryptorTransformer::decryptFrame(
   }
 
   if (!decryption_success) {
-    if (last_dec_error_ != FrameCryptionState::kDecryptionFailed) {
+    if (last_dec_error_ != FrameCryptionState::kDecryptionFailed || ratchet_count >= key_provider_->options().ratchet_window_size) {
```
Comment (theomonnom) [inline api/crypto/frame_crypto_transformer.cc]: Will this call multiple time the kDecryptionFailed callback?

File: api/crypto/frame_crypto_transformer.cc [hunk 2]
```diff
@@ -371,8 +388,9 @@ void FrameCryptorTransformer::encryptFrame(
 
   rtc::ArrayView<const uint8_t> date_in = frame->GetData();
   if (date_in.size() == 0 || !enabled_cryption) {
-    sink_callback->OnTransformedFrame(std::move(frame));
-    return;
+      RTC_LOG(LS_WARNING)
```
Comment (theomonnom) [inline api/crypto/frame_crypto_transformer.cc]: - Will you undo this change? (When FrameCryptor is not enabled, discard the input frame)

**Review by theomonnom — CHANGES_REQUESTED** [review]

**Review by theomonnom — APPROVED** [review]
otherwise. smtg

File: api/crypto/frame_crypto_transformer.cc [hunk 3]
```diff
@@ -490,25 +508,22 @@ void FrameCryptorTransformer::decryptFrame(
...
+      << "FrameCryptorTransformer::decryptFrame() date_in.size() == 0 || enabled_cryption == false";
     return;
   }
 
   auto uncrypted_magic_bytes = key_provider_->options().uncrypted_magic_bytes;
   if (uncrypted_magic_bytes.size() > 0 &&
-      date_in.size() >= uncrypted_magic_bytes.size() + 1) {
+      date_in.size() >= uncrypted_magic_bytes.size()) {
     auto tmp =
-        date_in.subview(date_in.size() - (uncrypted_magic_bytes.size() + 1),
```
Comment (theomonnom) [inline api/crypto/frame_crypto_transformer.cc:501]: is it the only change made to sif?

File: api/crypto/frame_crypto_transformer.cc (hunk 2 above)
Comment (cloudwebrtc) [reply api/crypto/frame_crypto_transformer.cc]: yes, I should revert this change

File: api/crypto/frame_crypto_transformer.cc (hunk 1 above)
Comment (cloudwebrtc) [reply api/crypto/frame_crypto_transformer.cc]: There seems to be no need to judge ratchet_count here, because when decryption_success == false, either ratchet_window_size = 0, or ratchet_count exceeds the number of times，
change to this should be ok
```cpp
  if (!decryption_success) {
      if(key_handler->DecryptionFailure()) {
        if (last_dec_error_ != FrameCryptionState::kDecryptionFailed) {
          last_dec_error_ = FrameCryptionState::kDecryptionFailed;
          onFrameCryptionStateChanged(last_dec_error_);
        }
      }
    return;
  }
```

File: api/crypto/frame_crypto_transformer.cc (hunk 3 above)
Comment (cloudwebrtc) [reply api/crypto/frame_crypto_transformer.cc:501]: Yes, when we discussed with @lukasIO  earlier, we reserved a 1 byte for frameType. It seems that it has been abandoned in the latest js-sdk.

Comment (lukasIO) [reply api/crypto/frame_crypto_transformer.cc:501]: yeah sorry, we decided on that when discussing with Raja, should have made this more explicit!
